### ISSUE TITLE
Computing angle between two rotations

### ISSUE THREAD

**Issue by kaufManu** [moble/quaternion#58]
I have two rotation matrices and I want to calculate the angle required to rotate one matrix into the other. To do so, I convert the two rotation matrices into quaternions `q1` and `q2`, compute `q = inv(q1) * q2` and then compute the norm of the vector that results when converting `q` into a rotation vector. With this implementation - while it technically returns a correct result - I noticed that the computed angle does not correspond to the shortest possible rotation. To check this, I also compute the angle of rotation with OpenCV's `Rodrigues` formula (see following code snippet which is a bit lengthy but self-contained).

```
import numpy as np
import quaternion
import cv2


def rad2deg(v):
    """
    Convert from radians to degrees.
    """
    return v * 180.0 / np.pi


def conj(qs):
    """Computes the conjugate of the quaternions, i.e. negates the imaginary parts."""
    return np.conjugate(qs)


def norm(qs):
    """Computes the length of the quaternions, i.e. the sum of the squares of the real and imaginary parts."""
    qs_r = np.reshape(qs, [-1])
    qs_arr = quaternion.as_float_array(qs_r)
    norms = np.sum(qs_arr * qs_arr, axis=-1)
    return np.reshape(norms, qs.shape)


def invert(qs):
    """Inverts the quaternions, i.e. returns the normalized conjugates."""
    conj_ = quaternion.as_float_array(conj(qs))
    normalized = conj_ / norm(qs)[..., np.newaxis]
    return quaternion.as_quat_array(normalized)


def Rx(angle):
    return np.array([[1.0, 0.0, 0.0],
                     [0.0, np.cos(angle), -np.sin(angle)],
                     [0.0, np.sin(angle), np.cos(angle)]])

def Ry(angle):
    return np.array([[np.cos(angle), 0.0, np.sin(angle)],
                     [0.0, 1.0, 0.0],
                     [-np.sin(angle), 0.0, np.cos(angle)]])

def Rz(angle):
    return np.array([[np.cos(angle), -np.sin(angle), 0.0],
                     [np.sin(angle), np.cos(angle), 0.0 ],
                     [0.0, 0.0, 1.0]])

def angle_error_q(rots1, rots2):
    q1 = quaternion.from_rotation_matrix(rots1)
    q2 = quaternion.from_rotation_matrix(rots2)

    q = invert(q1) * q2
    angles = np.linalg.norm(quaternion.as_rotation_vector(q), axis=-1)

    return rad2deg(angles)


def angle_error_rodrigues(rots1, rots2):
    rots2t = np.transpose(rots2, [0, 2, 1])
    E = np.matmul(rots1, rots2t)

    angles = []
    for i in range(E.shape[0]):
        r, _ = cv2.Rodrigues(E[i])
        angles.append(np.linalg.norm(r))
    return rad2deg(np.array(angles))


rots1 = np.array([Rx(np.pi),
                  np.matmul(Ry(np.pi), Rx(np.pi)),
                  np.matmul(Rz(np.pi), Ry(np.pi)),
                  Rx(np.pi/2),
                  Ry(np.pi/4),
                  Rz(np.pi/8)])

rots2 = np.array([Rx(np.pi/2),
                  Ry(np.pi/2),
                  np.matmul(Rz(np.pi), Ry(np.pi)),
                  Rx(-np.pi),
                  Ry(-np.pi),
                  Rz(-np.pi)])

print('with quaternions: ', angle_error_q(rots1, rots2))
print('with Rodrigues:   ', angle_error_rodrigues(rots1, rots2))
```

This prints
```
with quaternions:  [ 270.   180.     0.   270.   225.   202.5]
with Rodrigues:    [  90.   180.     0.    90.   135.   157.5]
```

Is this expected behaviour? Or is may be my implementation of the inverse false? As I said, the output is technically correct, however for my purposes I need the shortest possible angle of rotation as I need the value for an error metric.

It might also have something to do with the following observation. To find out more about what was going on, I installed [this other quaternion library](http://kieranwynn.github.io/pyquaternion/). I then construct a quaternion from the same rotation matrix with both your and this other package and print the result:

```
import numpy as np
import quaternion
from pyquaternion import Quaternion

Q1 = Quaternion(matrix=Rx(np.pi))
q1 = quaternion.from_rotation_matrix(Rx(np.pi))

print(Q1)
print(quaternion.as_float_array(q1)[0])
```

This prints
```
0.000 +1.000i +0.000j +0.000k
[ 0. -1. -0. -0.]
```

So, may be you are assuming a different convention in your implementation? If so, what do I have to do to retrieve the shortest angle of rotation (i.e. essentially get the same result as with the Rodrigues formula)?

**Comment by moble** [comment]
**TL;DR**: That's the correct behavior.  Use `quaternion.rotation_intrinsic_distance(q1, q2)` to get the angle between two __rotations__, instead of between two quaternions.  (But try to remember that while quaternions uniquely specify rotations, rotations do not uniquely specify quaternions.)

You are running into the fact that quaternions are ["spinorial"](https://en.wikipedia.org/wiki/Spinor), which means — for example — that a quaternion that rotates by 90° is *not the same* as a quaternion that rotates about the same axis by -270°, or about the opposite axis by +270°.  Similarly (but perhaps more surprisingly), a quaternion that rotates by 360° is not the same as a quaternion that rotates by 0°; you have to go through a full 720° to get back to where you started in quaternion space.

Obviously, this is not the case with rotation matrices; they are not spinorial.  But this means that there is not a unique quaternion to represent any given rotation matrix; there are actually two, and they just differ by a negative sign.  To see why, write down the formula to rotate a vector `v` by some quaternion `Q`:

    vprime = Q * v * Q.conj()

Notice that there are two places `Q` appears, so we can just flip its sign without affecting the result:

    vprime = (-Q) * v * (-Q).conj()

It turns out that when you rotate a quaternion by 360°, it gives you back precisely the negative of that quaternion.  In fact, technically you're going even farther because you're not just using the conjugate in the formulas above; you're actually normalizing your quaternions, which means that you could multiply the quaternion by *any* nonzero number (not just -1), and get the same rotation.  I'm assuming that the quaternions are already normalized, so they're technically ["rotors"](https://en.wikipedia.org/wiki/Rotor_(mathematics)), which is a term you see in several of the functions in this package.

So any time you ask "What quaternion will give me the same rotation as this rotation matrix?", there are actually *two* correct answers, which just have opposite signs.  Thus, when `pyquaternion` gives `Q1=[0, 1, 0, 0]` while `quaternion` gives `q1=[0, -1, 0, 0]` for your chosen rotation matrix, we're both right.  (And from what you've shown me at least, I don't see any reason to think that `pyquaternion` uses different conventions than `quaternion`.  But we evidently have different algorithms for choosing a quaternion based on a rotation matrix.)

You may wonder why any functions bother being sensitive to this sign difference when it doesn't make any difference *for a rotation*.  The answer is that it makes all the difference in the world when you're manipulating multiple quaternions.  For example, if you have a quaternion function of time and you want to differentiate it, you want the function to be pretty smooth.  But if, for example, my functions always chose the "smallest" equivalent rotation angle, the quaternion function would eventually get to a point where I would suddenly have to jump to a totally different quaternion (the negative), which would make the derivative totally crazy.  Same deal with interpolation by `squad` and `slerp`; a physical object would barely rotate between those two time steps, but the interpolated quaternion would suddenly spin through almost 360° for no reason.  So except in a few very specific cases, the functions in any good quaternion package have to respect those differences.

Finally, you may say "That's all well and good, but I just care about a few discrete rotations, and I just want the rotational angle, not the spinorial angle.  Isn't there a function that just gives me that?"  The answer is yes:

```python
def angle_error_q(rots1, rots2):
    q1 = quaternion.from_rotation_matrix(rots1)
    q2 = quaternion.from_rotation_matrix(rots2)

    angles = np.array([quaternion.rotation_intrinsic_distance(r1, r2)
                       for r1, r2 in zip(q1, q2)])

    return rad2deg(angles)
```

**Comment by kaufManu** [comment]
This is a great answer - thanks a lot for taking the time to explain that to me! This has helped me to understand quaternions and their use better. The longer I work with this library, the more I love it.
